libraries/extractor/src/main/java/androidx/media3/extractor/DtsUtil.java

```diff
@@ -223,8 +357,262 @@ public static int getDtsFrameSize(byte[] data) {
+    frameBits.skipBits( 32 + 8 + 2); // SYNC, UserDefinedBits, nExtSSIndex
+    // Unpack the num of bits to be used to read header size
+    int headerSizeType = frameBits.readBits(1);
+    if (headerSizeType == 0){
+      nuBits4Header = 8;
+    }
+    else{
+      nuBits4Header = 12;
+    }
+    // Unpack the substream header size
+    return frameBits.readBits(nuBits4Header) + 1;
+  }
+
+  /**
+   * Returns the DTS format given {@code data} containing the DTS-UHD(Profile 2) frame according to
+   * ETSI TS 103 491 V1.2.1 (2019-05), Section 6.4.3.
+   *
+   * @param frame The DTS-UHD frame to parse.
+   * @return The DTS format parsed from data in the header.
+   */
+  public static DtsFormatInfo parseDtsUhdFormat(byte[] frame) throws ParserException {
+    ParsableBitArray frameBits = getNormalizedFrameHeader(frame);
+    int sampleRate = 0;
+    int sampleCount = 0;
+    boolean bSyncFrameFlag = false;
+
+    int nSyncWord = frameBits.readBits(32);
+    if (nSyncWord == SYNC_VALUE_UHD_FTOC_SYNC_BE) {
+      bSyncFrameFlag = true;
+    }
+    int[] ucTable1 = new int[] { 5, 8, 10, 12 };
+    int nFTOCPayloadinBytes = extractVarLenBitFields(frameBits, ucTable1, true) + 1;
+
+    if (bSyncFrameFlag) {
+      // fullChannelBasedMixFlag, ETSI TS 103 491 V1.2.1, Section 6.4.6.1
+      if(frameBits.readBits(1) != 1) {
+        throw ParserException.createForMalformedContainer(
+            /* message= */ "Only supports full channel mask-based audio presentation",
+            /* cause= */ null);
+      }
+
+      int nBaseDurationIndex = frameBits.readBits(2);
+      sampleCount = BASE_DURATION[nBaseDurationIndex] * (frameBits.readBits(3) + 1);
+      int clockRateIndex = frameBits.readBits(2);
+      int clockRateHertz = 0;
+      switch (clockRateIndex) {
+        case 0:
+          clockRateHertz = 32000;
+          break;
+        case 1:
+          clockRateHertz = 44100;
+          break;
+        case 2:
+          clockRateHertz = 48000;
+          break;
+      }
+      // Read time stamp information
+      if (frameBits.readBits(1) == 1) {
+        frameBits.skipBits(32 + 4);
+      }
+      int samplRateMultiplier = (1 << frameBits.readBits(2));
+      sampleRate = clockRateHertz * samplRateMultiplier;
+    }
+
+    int chunkPayloadBytes = 0;
+    int numOfMDChunks = bSyncFrameFlag ? 1 : 0; // Metadata chunks
+    for (int nmdc = 0; nmdc < numOfMDChunks; nmdc++) {
+      int[] ucTable2 = new int[] { 6, 9, 12, 15 };
+      int nuMDChunkSize = extractVarLenBitFields(frameBits, ucTable2, true);
+      if (nuMDChunkSize > 32767) {
+        throw ParserException.createForMalformedContainer(
+            /* message= */ "Unsupported metadata chunk size in DTS UHD header: " + nuMDChunkSize,
+            /* cause= */ null);
+      }
+      chunkPayloadBytes += nuMDChunkSize;
+    }
+
+    // Ony one audio chunk is supported
+    int numAudioChunks = 1;
+    for (int nac = 0; nac < numAudioChunks; nac++) {
+      int acID = 256, nuAudioChunkSize;
```

Comment:
where does this 256 come from?

Reply:
256 is used as a number which indicates the invalid audio chunck ID. The valid range is up to 255 only.

Reply:
Why is it correct to use an invalid audio chunk ID here if? The spec (ETSI TS 103 491 V1.2.1 table 6-20) makes it look like we should look up the audio chunk ID of the previous sync frame? I think that's what this line is implying:

```
 // Get the audio chunk ID from the stored list
 unAudioChunkID =
m_CAudioChunkNaviLits.m_pCAudioChunksNaviParams[unListIndex].m_unAudioChunkID; 
```

Reply:
@icbaker, as you mentioned, in non-syn frames we need to use the chunk ID reported in the previous sync frame.
But I am trying to avoid keeping a state inside the utility function.
In any sync frame, the _unAudioChunkID_ extracted from the stream will be a non-zero value(as per the current bitstream stream construction, it will be 0x01). So we can consider that the subsequent non-sync frame will get the stored non-zero value for the _unAudioChunkID_.

Reply:
> But I am trying to avoid keeping a state inside the utility function.

I understand this is a nice goal, but doesn't the spec **require** keeping some form of state?

I'm afraid I'm unclear how the current implementation is compatible with the spec - since the spec says on non-sync frames we need to use the audio chunk ID of the previous sync frame, and the implementation says that non-sync frames have an invalid audio chunk ID.

Reply:
@icbaker , I made an update in the code to handle this. Could you check my last commit([e99a618](https://github.com/androidx/media/pull/275/commits/e99a6185d3f22e376243e1e29524766d7f700c43))?